server/src/test/java/org/elasticsearch/cluster/coordination/ReconfiguratorTests.java

```diff
@@ -0,0 +1,195 @@
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterState.VotingConfiguration;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptySet;
+import static org.elasticsearch.cluster.coordination.Reconfigurator.CLUSTER_MASTER_NODES_FAILURE_TOLERANCE;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class ReconfiguratorTests extends ESTestCase {
+
+    public void testReconfigurationExamples() {
+        for (int masterNodesFailureTolerance = 0; masterNodesFailureTolerance <= 2; masterNodesFailureTolerance++) {
+            check(nodes("a"), conf("a"), masterNodesFailureTolerance, conf("a"));
+            check(nodes("a", "b"), conf("a"), masterNodesFailureTolerance, conf("a"));
+            check(nodes("a", "b", "c"), conf("a"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c"), conf("a", "b"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d", "e"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c", "d", "e"));
+            check(nodes("a", "b"), conf("a", "b", "e"), masterNodesFailureTolerance,
+                masterNodesFailureTolerance == 1 ? conf("a", "b", "e") : conf("a"));
+            check(nodes("a", "b", "c"), conf("a", "b", "e"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d"), conf("a", "b", "e"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d", "e"), conf("a", "f", "g"), masterNodesFailureTolerance, conf("a", "b", "c", "d", "e"));
+            check(nodes("a", "b", "c", "d"), conf("a", "b", "c", "d", "e"), masterNodesFailureTolerance,
+                masterNodesFailureTolerance == 2 ? conf("a", "b", "c", "d", "e") : conf("a", "b", "c"));
+        }
+
+        // Retiring a single node shifts the votes elsewhere if possible.
+        check(nodes("a", "b"), retired("a"), conf("a"), 0, conf("b"));
+
+        // If the safety level was never reached then retirement can take place
+        check(nodes("a", "b"), retired("a"), conf("a"), 1, conf("b"));
+        check(nodes("a", "b"), retired("a"), conf("b"), 1, conf("b"));
```

Comment:
these 2 tests don't make sense anymore? We never reconfigure that way. One more reason not to hide the `config.getNodeIds().size() < 2 * masterNodesFailureTolerance + 1)` case within `check`